Gems/ROS2/Code/Source/Gripper/GripperActionServer.cpp

```diff
@@ -0,0 +1,129 @@
+    {
+        if (!IsReadyForExecution())
+        {
+            AZ_Printf("GripperActionServer", "GripperActionServer::handleGoal: Rejected goal, already executing\n");
+            if (m_goalHandle)
+            {
+                AZ_Trace(
+                    "GripperActionServer",
+                    " is_active: %d,  is_executing %d, is_canceling : %d\n",
+                    m_goalHandle->is_active(),
+                    m_goalHandle->is_executing(),
+                    m_goalHandle->is_canceling());
+            }
+            return rclcpp_action::GoalResponse::REJECT;
+        }
+
+        AZ::Outcome<void, AZStd::string> commandOutcome = AZ::Failure(AZStd::string("No gripper component found!"));
+        GripperRequestBus::EventResult(
+            commandOutcome, m_entityId, &GripperRequestBus::Events::GripperCommand, goal->command.position, goal->command.max_effort);
+        if (!commandOutcome.IsSuccess())
+        {
+            AZ_Trace("GripperActionServer", "GripperCommand could not be accepted: %s\n", commandOutcome.GetError().c_str());
+            return rclcpp_action::GoalResponse::REJECT;
+        }
+        AZ_Trace("GripperActionServer", "GripperActionServer::handleGoal: GripperCommand accepted!\n");
+        return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
+    }
+
+    rclcpp_action::CancelResponse GripperActionServer::GoalCancelledCallback(const std::shared_ptr<GoalHandleGripperCommand> goal_handle)
+    {
+        AZ::Outcome<void, AZStd::string> cancelOutcome = AZ::Failure(AZStd::string("No gripper component found!"));
+        GripperRequestBus::EventResult(cancelOutcome, m_entityId, &GripperRequestBus::Events::CancelGripperCommand);
+
+        if (!cancelOutcome)
+        { // This will not happen in a simulation unless intentionally done for behavior validation
+            AZ_Trace("GripperActionServer", "Cancelling could not be accepted: %s\n", cancelOutcome.GetError().c_str());
+            return rclcpp_action::CancelResponse::REJECT;
+        }
+        return rclcpp_action::CancelResponse::ACCEPT;
+    }
+
+    void GripperActionServer::GoalAcceptedCallback(const std::shared_ptr<GoalHandleGripperCommand> goal_handle)
+    {
+        AZ_Trace("GripperActionServer", "Goal accepted!\n");
+        m_goalHandle = goal_handle;
+    }
+
+    void GripperActionServer::CancelGoal(std::shared_ptr<GripperCommand::Result> result)
+    {
+        AZ_Assert(m_goalHandle, "Invalid goal handle!");
+        if (m_goalHandle && m_goalHandle->is_canceling())
+        {
+            AZ_Trace("FollowJointTrajectoryActionServer", "Cancelling goal\n");
+            m_goalHandle->canceled(result);
+        }
+    }
+
+    void GripperActionServer::GoalSuccess(std::shared_ptr<GripperCommand::Result> result)
+    {
+        AZ_Assert(m_goalHandle, "Invalid goal handle!");
+        if (m_goalHandle && m_goalHandle->is_executing())
```

Comment:
A check for the is_cancelling should also be added. It is possible for the action server to succeed when in the canceling state.

Reply:
There's a check if it's executing, a goal cannot be simultaneously cancelling and executing:
https://github.com/ros2/rclcpp/blob/humble/rclcpp_action/src/server_goal_handle.cpp